File: src/main/java/com/bingchunmoli/api/weather/task/WeatherNotionTask.java
package com.bingchunmoli.api.weather.task;

import com.bingchunmoli.api.exception.ApiTaskException;
import com.bingchunmoli.api.utils.SendMailUtil;
import com.bingchunmoli.api.weather.bean.WeatherDailyBean;
import com.bingchunmoli.api.weather.bean.WeatherSub;
import com.bingchunmoli.api.weather.bean.enums.Code;
import com.bingchunmoli.api.weather.service.WeatherService;
import jakarta.mail.MessagingException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Async;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.thymeleaf.context.Context;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author bingchunmoli
 */
@Slf4j
@Component
@ConditionalOnProperty(prefix = "moli.features.tasks", name = "weather-notification", havingValue = "true")
@RequiredArgsConstructor
public class WeatherNotionTask {
    private final WeatherService weatherService;
    private final ObjectMapper om;
    private final SendMailUtil mailUtil;


    @Scheduled(cron = "0 30 18 * * ?")
    public void notion() {
        final List<WeatherSub> list = weatherService.list();
        final Map<String, WeatherDailyBean> notifiedLocationMap = getNotifiedLocation(list);
        notionMessage(notifiedLocationMap, list);
    }

    @Async
    public void notionMessage(
            final Map<String, WeatherDailyBean> notifiedLocationMap,
            final List<WeatherSub> list
    ) {
        final Map<String, List<WeatherSub>> messageMap = list.stream()
                .collect(Collectors.groupingBy(WeatherSub::getLocation));
        for (final Map.Entry<String, WeatherDailyBean> entry : notifiedLocationMap.entrySet()) {
            final List<WeatherSub> weatherSub = messageMap.get(entry.getKey());
            final WeatherDailyBean value = entry.getValue();
            if (log.isDebugEnabled()) {
                log.debug("key: {}, message: {}", entry.getKey(), entry.getValue());
            }
            final Context context = new Context();
            weatherSub.forEach(v -> {
                context.setLocale(Locale.CHINA);
                context.setVariable("email", v.getEmail());
                context.setVariable("location", v.getLocation());
                context.setVariable("updateTime", value.getUpdateTime());
                context.setVariable("fxLink", value.getFxLink());
                context.setVariable("data", value.getDaily());
                try {
                    mailUtil.sendHtmlMail(v.getEmail(), "天气不好,记得带伞呦😘😘😘", "WeatherNotion", context);
                } catch (MessagingException e) {
                    log.error("发送邮件异常: ", e);
                    throw new ApiTaskException(e);
                }
            });
        }
    }

    private Map<String, WeatherDailyBean> getNotifiedLocation(final List<WeatherSub> list) {
        final List<String> locationList = list.stream().map(WeatherSub::getLocation).distinct().toList();
        final HashMap<String, WeatherDailyBean> map = new HashMap<>();
        for (final String location : locationList) {
            final JsonNode weather = weatherService.getWeatherByDay(3, location);
            final WeatherDailyBean weatherDailyBean;
            try {
                if (log.isDebugEnabled()) {
                    log.debug("location: {}, weather: {}", location, weather);
                }
                weatherDailyBean = om.treeToValue(weather, WeatherDailyBean.class);
            } catch (JacksonException e) {
                throw new ApiTaskException(e);
            }
            if (!Code.OK.getCode().equalsIgnoreCase(weatherDailyBean.getCode())) {
                if (log.isInfoEnabled()) {
                    log.info("locationList: {}, weatherDailyBean: {}", locationList, weatherDailyBean);
                }
            }
            if (log.isDebugEnabled()) {
                log.debug("weatherDailyBean: {}", weatherDailyBean);
            }
            if (weatherDailyBean.getDaily().stream().anyMatch(v -> Integer.parseInt(v.getIconDay()) >= 300)) {
                map.put(location, weatherDailyBean);
            }
        }
        if (log.isDebugEnabled()) {
            log.debug("WeatherDailyMap: {}", map);
        }
        return map;
    }

}

File: src/main/java/com/bingchunmoli/api/weather/task/WeatherNotionTask.kt
package com.bingchunmoli.api.weather.task

import com.bingchunmoli.api.utils.LogUtil.debug
import com.bingchunmoli.api.utils.LogUtil.info
import com.bingchunmoli.api.weather.bean.WeatherDailyBean
import com.bingchunmoli.api.weather.bean.enums.Code
import com.bingchunmoli.api.weather.service.WeatherService
import com.fasterxml.jackson.databind.ObjectMapper
import lombok.Data
import lombok.RequiredArgsConstructor
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

@Component
@RequiredArgsConstructor
class WeatherNotionTask (private val weatherService: WeatherService, private val om : ObjectMapper) {

    @Data
    class obj{
        var obj : Array<WeatherDailyBean>? = null
    }

    @Scheduled(cron = "0 30 18 * * ?")
    fun notion(){
        val list = weatherService.list()
        val locationList = list.stream().map { it.location }.distinct().toList()
        val map = HashMap<String, List<WeatherDailyBean.DailyBean>>()
        for (location in locationList) {
            val weather = weatherService.getWeatherByDay(3, location)
            info("weather: {}", weather)
            val weatherDailyBean = om.readValue(weather, obj::class.java)
            if (Code.OK.code != weatherDailyBean.obj?.first()?.code?.code) {
                info("locationList: {}, weatherDailyBean: {}", locationList, weatherDailyBean)
            }
            // FIXME: let null map {}
            val let = weatherDailyBean.obj?.first()?.daily?.let { map.put(location, it) }
        }
        debug("WeatherDailyMap: {}", map)
        val filter = map.filter { (_, v) -> v.first().iconDay.toInt() >= 300 }
        info("filter: {}, map: {}",filter, map)
//        TODO 通知
        TODO("通知")
    }
}